rs/consensus/src/dkg/payload_validator.rs

```diff
@@ -111,141 +119,159 @@ pub(crate) fn validate_payload(
         // chain.
         .expect("No DKG start block found for the parent block.");
     let last_dkg_summary = &last_summary_block.payload.as_ref().as_summary().dkg;
 
-    let current_height = parent.height.increment();
     let is_dkg_start_height = last_dkg_summary.get_next_start_height() == current_height;
 
-    if payload.is_summary() {
-        if !is_dkg_start_height {
-            return Err(InvalidDkgPayloadReason::DkgSummaryAtNonStartHeight(current_height).into());
+    match payload {
+        BlockPayload::Summary(summary_payload) => {
+            if !is_dkg_start_height {
+                return Err(
+                    InvalidDkgPayloadReason::DkgSummaryAtNonStartHeight(current_height).into(),
+                );
+            }
+            let expected_summary = payload_builder::create_summary_payload(
+                subnet_id,
+                registry_client,
+                crypto,
+                pool_reader,
+                last_dkg_summary,
+                &parent,
+                registry_version,
+                state_manager,
+                validation_context,
+                ic_logger::replica_logger::no_op_logger(),
+            )?;
+            if summary_payload.dkg != expected_summary {
+                return Err(InvalidDkgPayloadReason::MismatchedDkgSummary(
+                    expected_summary,
+                    summary_payload.dkg.clone(),
+                )
+                .into());
+            }
+            Ok(())
         }
-        let registry_version = pool_reader
-            .registry_version(current_height)
-            .expect("Couldn't get the registry version.");
-        let expected_summary = payload_builder::create_summary_payload(
-            subnet_id,
-            registry_client,
-            crypto,
-            pool_reader,
-            last_dkg_summary,
-            &parent,
-            registry_version,
-            state_manager,
-            validation_context,
-            ic_logger::replica_logger::no_op_logger(),
-        )?;
-        if payload.as_summary().dkg != expected_summary {
-            return Err(InvalidDkgPayloadReason::MismatchedDkgSummary(
-                expected_summary,
-                payload.as_summary().dkg.clone(),
+        BlockPayload::Data(data_payload) => {
+            if is_dkg_start_height {
+                return Err(
+                    InvalidDkgPayloadReason::DkgDealingAtStartHeight(current_height).into(),
+                );
+            }
+            let max_dealings_per_block = registry_client
+                .get_dkg_dealings_per_block(subnet_id, registry_version)
+                .map_err(DkgPayloadValidationFailure::FailedToGetMaxDealingsPerBlock)?
+                .unwrap_or_else(|| {
+                    panic!(
+                        "No subnet record found for registry version={} and subnet_id={}",
+                        registry_version, subnet_id
+                    )
```

Comment:
Yes, I don't think there is anything better we can do here. If the subnet record is empty, we wouldn't even enter this function as we would have panicked already [here](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/consensus/src/consensus/validator.rs?L981) because of [this check](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/consensus/utils/src/lib.rs?L306-311)